### PR TITLE
Vectorization of xrd_simulator

### PR DESCRIPTION
**Pull request by Marcraven**
This pull request is a generealized vectorization of most  procedures in xrd_simulator.

The main idea behind is to increase computation speed an facilitate code parallelization whenever possible. The motivation behind was my intention to use this code to simulate powder samples with several million tetrahedra.

The main files affected are:

# mesh.py

Once the mesh is created by meshio/pygalmesh, subsequent computations related to bounding spheres, centroids etc. have all been vectorized.

# polycrystal.py

The _diffract method has been extensively modified to compute all diffraction points simultaneously for each phase present in the sample. There is still work to be done on the side of the tetrahedron-beam intersection, in which a faster bounding-box approximation has been implemented. The convex-hull calculation for the ScatteringUnit is still a bottleneck, but so far I have no idea how to improve it.

Another minor improvement has been adding zd and yd to the scattering unit to prevent repeating the calculation later in the rendering.

# utils.py

Geometric calculations for circumspheres have been added here which help accelerate the mesh.py script a lot. A printvars utility has been added to help debugging memory issues during code execution.

# Current bottlenecks

- The tetrahedra-beam intersection
- The convex-hull calculation
- The intensity calculation during rendering

Vectorization of these last steps is not trivial and has to be thought through. The last one is probably doable although I could not find the time.

# Suggestions

- Include beam absorption physics
- Allow for a range of wavelengths to be simulated rather than a single one
- Do not impose >0 rotation for rigid body motions, perhaps one only wants to translate.

### CODE REVIEW

**Comment by AxelHenningsson** [comment]
Did these changes introduce new dependecies: i see "ModuleNotFoundError: No module named 'pandas'" in the logs.

**Comment by Marcraven** [comment]
> Did these changes introduce new dependecies: i see "ModuleNotFoundError: No module named 'pandas'" in the logs.

Hi Axel, yes pandas is now used briefly in polycrystal to make the information more readable in tables. I could be done using purely numpy, but it is quite hard to debug a matrix with 17 columns being each column a different concept...

**Comment by Marcraven** [comment]
On the other hand there is a conflict in the test for arm64 and python 3.10.0 not being available for that architecture. Could the test be changed to Python 3.10.11 or some other build which is available for arm64 at https://raw.githubusercontent.com/actions/python-versions/main/versions-manifest.json?

**Comment by AxelHenningsson** [comment]
Hi again,

Yeah, 3.10.11 should work.... but there is a recent issue with macos-latest changing image from 13 to 14. I belive conda is more or less broken for now.... [see this issue](https://github.com/s-weigand/setup-conda/issues/432) :laughing: 

I have pruned the action set to just run all tests on a linux machine for now. Hopefully we can improve on this in the future....

That being said there seem to be [tests failing here](https://github.com/FABLE-3DXRD/xrd_simulator/actions/runs/8894015914/job/24421384604?pr=9). Did you run a pytest on you local machine before opening this pull-request?

Cheers :slightly_smiling_face: 
Axel

**Comment by AxelHenningsson** [comment]
@Marcraven Could you add in the latest action changes into this request and we will see if the tests passes now.

Cheers
Axel

**Comment by Marcraven** [comment]
In this pull request ScatteringUnit has two new arguments, zd and yd.

That is because in a normal pipeline, the intersection between scattering unit and detector is done twice, one to filter out the units that do not intersect and once to render.

This way, zd and yd can be passed from the first calculation to the second one to save computation time.

I am now re-writing test_detector to adapt it to this new configuration, but this is all open for discussion.

**Comment by Marcraven** [comment]
**test_laue.py** fails without using any new part from my code. The strain tensor generated in get_pseudorandom_crystal and passed to _epsilon_to_b is not a tensor, but a vector. When given a tensor, then _epsilon_to_b fails because **F = np.linalg.cholesky(C).transpose(0,2,1)** cannot transpose in 3D a matrix that only has 2D.

**Comment by AxelHenningsson** [comment]
On it. :)

**Comment by AxelHenningsson** [comment]
Hi @Marcraven,

I'm very happy to see you working on making the test pass. I'm looking forward to merging all of this.

Regarding your question, _epsilon_to_b is indeed supposed to take a vector of six values. This is why the call to strain_tensor = _strain_as_tensor(epsilon) at the start of the function is necessary—it converts the vector into a strain tensor.

I’m having trouble understanding the problem you’re encountering. The tests pass fine on my local Linux machine when I clone the current repo and install everything.

Could it be that you’re running a new version of utils._epsilon_to_b? If you have made recent changes, there might be discrepancies between our environments.

Let’s sync up to resolve this!

Cheers,
Axel

**Comment by AxelHenningsson** [comment]
Your pull request confirms [this suspicion](https://github.com/FABLE-3DXRD/xrd_simulator/pull/9/files#diff-986bd88c118aadb928e6ba5bb9380186e03bc8028140353c2f4bb15822ee6d41) with `F = np.linalg.cholesky(C).transpose(0,2,1)` being newly added code :)

**Comment by Marcraven** [comment]
I am sorry! this is what happens when you come back to your code after weeks... 

I modified _epsilon_to_b to accept any number of crystal strains, i will make it work for 2D now...

**Comment by Marcraven** [comment]
> In this pull request ScatteringUnit has two new arguments, zd and yd.
> 
> That is because in a normal pipeline, the intersection between scattering unit and detector is done twice, one to filter out the units that do not intersect and once to render.
> 
> This way, zd and yd can be passed from the first calculation to the second one to save computation time.
> 
> I am now re-writing test_detector to adapt it to this new configuration, but this is all open for discussion.

What do you think of this @AxelHenningsson ?

**Comment by AxelHenningsson** [comment]
Fantastic! No worries at all—if anything, I’m at fault for being slow to reply to the excellent work you’re doing!

Regarding the `ScatteringUnit`, that sounds great. Since the `ScatteringUnit` is primarily used internally within `xrd_simulator`, complexity is less of a concern compared to components that interface directly with the user. Good catch!

Cheers,
Axel

**Comment by Marcraven** [comment]
*_instantiate_eB* in now iterates over phases instead of iterating over elements. This is much more eficient because the loop runs less than 10 times typically, instead of whatever number of tets we have. But in order to work like this B0 has to be computed in _instantiate_eB and passed down to lab_strain_to_B_matrix and then to _epsilon_to_B.

**Comment by AxelHenningsson** [comment]
Sounds great.

Are tests getting close to passing yet?

Cheers
Axel

**Comment by Marcraven** [comment]
Hi @AxelHenningsson ,

test_beam, test_detector, test_laue, test_mesh, test_motion, are passing already. Most times the problem is related to dimensional mismatchs, annoying but not fundamental.

I will keep you posted.

**Comment by AxelHenningsson** [comment]
Sounds like good progress! 👌

**Comment by Marcraven** [comment]
This is ready to ship. The docstrings should still be verified to fit with current changes though.

The key issue limiting performance right now are convexhull operations. I dont know any vectorized way to intersect many tetrahedra with some polyhedra simultaneously, probably there is some GPU way of doing it.

File: xrd_simulator/detector.py [hunk 1]
```diff
@@ -107,7 +107,10 @@ def render(
             frames_to_render (:obj:`int` or :obj:`iterable` of :obj:`int` or :obj:`str`): Indices of the frame in the :obj:`frames` list
                 to be rendered. Optionally the keyword string 'all' can be passed to render all frames of the detector.
             lorentz (:obj:`bool`): Weight scattered intensity by Lorentz factor. Defaults to False.
-            polarization (:obj:`bool`): Weight scattered intensity by Polarization factor. Defaults to False.
+            polarization (:obj:`bool`): Weight scattereeen missing from the libraries of natural science experimentalists. Created and edited by a number of fellow scientists, including myself, who either work at neutron sources, operate neutron imaging instrumentation or are users of these facilities.
```
Comment (AxelHenningsson) [inline xrd_simulator/detector.py:110]: 😆 ?!

File: xrd_simulator/scattering_unit.py [hunk 2]
```diff
@@ -33,12 +33,11 @@ class ScatteringUnit(object):
         incident_polarization_vector (:obj:`numpy array`): Unit vector of linear polarization ```shape=(3,)```
         rotation_axis (:obj:`numpy array`): Sample motion rotation axis ```shape=(3,)```
         time (:obj:`numpy array`): Parametric value in range [0,1] determining at which instance during sample
-            motion the scattering occured.
+            motion the scattering ozd2,ydccured.
```
Comment (AxelHenningsson) [inline xrd_simulator/scattering_unit.py:36]: 😆 ?!

File: xrd_simulator/utils.py [hunk 3]
```diff
@@ -162,7 +163,7 @@ def alpha_to_quarternion(alpha_1, alpha_2, alpha_3):
 def lab_strain_to_B_matrix(
```
Comment (AxelHenningsson) [inline xrd_simulator/utils.py:163]: Docstrings need to be modified.

File: xrd_simulator/utils.py [hunk 4]
```diff
@@ -317,3 +314,78 @@ def get_misorientations(orientations):
         difference_rotation = Rotation.from_matrix(U.dot(mean_orientation.T))
         misorientations[i] = Rotation.magnitude(difference_rotation)
     return misorientations
+
+
+def compute_sides(points):
```
Comment (AxelHenningsson) [inline xrd_simulator/utils.py:319]: to be hidden.

File: xrd_simulator/utils.py [hunk 5]
```diff
@@ -317,3 +314,78 @@ def get_misorientations(orientations):
...
+    # Compute the distances by taking the square root of the squared distances
+    dist_mat = np.sqrt(squared_distances)
+    
+    # Extract the 1-to-1 values from the distance matrix
+    distances = np.hstack((dist_mat[:,0,1:],dist_mat[:,1,2:],dist_mat[:,2,3][:,np.newaxis]))
+    
+    return distances
+
+
+def circumsphere_of_segments(segments):
```
Comment (AxelHenningsson) [inline xrd_simulator/utils.py:340]: to be hidden

File: xrd_simulator/utils.py [hunk 6]
```diff
@@ -317,3 +314,78 @@ def get_misorientations(orientations):
...
+    """Computes the minimum circumsphere of n segments given by a numpy array of vertices nx2x3"""
+    centers = np.mean(segments,axis=1)
+    radii = np.linalg.norm(centers-segments[:,0,:],axis=1)
+    return centers, radii
+
+def circumsphere_of_triangles(triangles):
```
Comment (AxelHenningsson) [inline xrd_simulator/utils.py:346]: to be hidden

File: xrd_simulator/utils.py [hunk 7]
```diff
@@ -317,3 +314,78 @@ def get_misorientations(orientations):
...
+    norm_abXac = np.linalg.norm(abXac,axis=1)
+    
+    a_to_centre = (np.cross(abXac,ab)*((np.linalg.norm(ac,axis=1)**2)[:,np.newaxis])+np.cross(acXab,ac)*((np.linalg.norm(ab,axis=1)**2)[:,np.newaxis]))/(2*(np.linalg.norm(abXac,axis=1)**2)[:,np.newaxis])
+
+    centers = triangles[:,0,:]+ a_to_centre    
+    radii = np.linalg.norm(a_to_centre,axis=1)
+    
+    return centers, radii
+
+def circumsphere_of_tetrahedrons(tetrahedra):
```
Comment (AxelHenningsson) [inline xrd_simulator/utils.py:363]: to be hidden

File: xrd_simulator/utils.py [hunk 8]
```diff
@@ -317,3 +314,78 @@ def get_misorientations(orientations):
...
+                  np.linalg.norm(v2,axis=1)**2-np.linalg.norm(v0,axis=1)**2,
+                  np.linalg.norm(v3,axis=1)**2-np.linalg.norm(v0,axis=1)**2)).T
+
+    centers = np.matmul(np.linalg.inv(A),B[:,:,np.newaxis])[:,:,0]
+    radii = np.linalg.norm((tetrahedra.transpose(2,0,1)-centers.transpose(1,0)[:,:,np.newaxis])[:,:,0],axis=0)
+    
+    return centers, radii
+
+
+def sizeof_fmt(num, suffix='B'):
```
Comment (AxelHenningsson) [inline xrd_simulator/utils.py:381]: to be hidden

File: xrd_simulator/utils.py [hunk 9]
```diff
@@ -317,3 +314,78 @@ def get_misorientations(orientations):
...
+    ''' by Fred Cirera,  https://stackoverflow.com/a/1094933/1870254, modified'''
+    for unit in ['','Ki','Mi','Gi','Ti','Pi','Ei','Zi']:
+        if abs(num) < 1024.0:
+            return "%3.1f %s%s" % (num, unit, suffix)
+        num /= 1024.0
+    return "%.1f %s%s" % (num, 'Yi', suffix)
+
+def printvars(vars):
```
Comment (AxelHenningsson) [inline xrd_simulator/utils.py:389]: to be hidden

**Comment by AxelHenningsson** [comment]
Fantastic! This looks good to me in general! 🙂 

A corner case for the mesh spheres came up in my testing on your fork - it is related to loss of precision in the line segment and triangle sphere wrapper functions. The fix is just to expand the sphere radii very slightly to ensure that the triangle and line segments are actually containing the nodes they claim to contain. Without this I get bounding sphere radii that change as a function of pure rigid body motions which makes testing awkward. see commit d145326

I think we should add docstrings to the functions that are user exposed before we can merge this. It would be a shame to have a bunch of docs that are lying about the code. The starting point would be to hide the functions that are not supposed to be used by the user using the underscore notation `_hidden_function`. For the remaining functions that changed the docs should follow the same sphinx format as before, this will enable the docs to get autogenerated. It looks like you have already added docstrings in some places?

Cheers
Axel

**Comment by AxelHenningsson** [comment]
@Marcraven : Please let me know if you will have the possibility to get the docs for this major  pull request in shape (don't need to be perfect). It would be a shame to loose all the good progress. ❤️  

Once this gets merged I think we deserve to roll up the version count and move from beta into alpha with a new releases on pypi and conda channels.

Cheers
Axel

**Comment by Marcraven** [comment]
Hi Axel,

I am sorry, I want to do it this week. Hopefully tomorrow!

Un saludo

---
Marc Raventós Tato

**Comment by Marcraven** [comment]
> The fix is just to expand the sphere radii very slightly to ensure that the triangle and line segments are actually containing the nodes they claim to contain.

I increased the radii by 0.1% in the cases of line and triangle segments. ¿Would that be enough?

**Comment by AxelHenningsson** [comment]
> > The fix is just to expand the sphere radii very slightly to ensure that the triangle and line segments are actually containing the nodes they claim to contain.
> 
> I increased the radii by 0.1% in the cases of line and triangle segments. ¿Would that be enough?

Sounds good!

**Pull request merged**